capsules/src/hmac.rs

```diff
@@ -447,39 +451,39 @@ impl<'a, H: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> LegacyDri
                             if app.pending_run_app.is_some() {
                                 // No more room in the queue, nowhere to store this
                                 // request.
-                                ReturnCode::ENOMEM
+                                CommandResult::failure(ErrorCode::NOMEM)
                             } else {
                                 // We can store this, so lets do it.
                                 app.pending_run_app = Some(appid);
-                                ReturnCode::SUCCESS
+                                CommandResult::success()
                             }
                         })
                         .unwrap_or_else(|err| err.into())
                 }
             }
 
             // default
-            _ => ReturnCode::ENOSUPPORT,
+            _ => CommandResult::failure(ErrorCode::NOSUPPORT),
         }
     }
 }
 
 pub struct App {
-    callback: OptionalCell<Callback>,
+    callback: Callback,
     pending_run_app: Option<AppId>,
-    key: Option<AppSlice<SharedReadWrite, u8>>,
-    data: Option<AppSlice<SharedReadWrite, u8>>,
-    dest: Option<AppSlice<SharedReadWrite, u8>>,
+    key: ReadWriteAppSlice,
+    data: ReadWriteAppSlice,
+    dest: ReadWriteAppSlice,
 }
 
 impl Default for App {
     fn default() -> App {
         App {
-            callback: OptionalCell::empty(),
+            callback: Callback::default(),
             pending_run_app: None,
-            key: None,
-            data: None,
-            dest: None,
+            key: ReadWriteAppSlice::default(),
+            data: ReadWriteAppSlice::default(),
+            dest: ReadWriteAppSlice::default(),
         }
     }
 }
```

Comment:
I think this could just derive `Default`, which would cut some unneeded lines of code.